File: penguin-ng/src/app/penguin/service/penguin.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from "@angular/common/http";
import { Observable } from "rxjs";
import { Penguins } from "../model/penguins";
import { Penguin } from "../model/penguin";
import { NewPenguin } from "../model/new-penguin";
import { ImagesInfo } from "../model/images-info";

/**
 * Service responsible for managing penguins.
 */
@Injectable({
  providedIn: 'root'
})
export class PenguinService {

  /**
   * @param httpClient http client
   */
  constructor(private httpClient: HttpClient) { }

  /**
   * Fetches all penguins
   */
  getPenguins(): Observable<Penguins> {
    return this.httpClient.get<Penguins>(`/api/penguins`);
  }

  /**
   * Fetches penguin with given taxon key.
   *
   * @param taxonKey taxon key
   */
  getPenguin(taxonKey: string): Observable<Penguin> {
    return this.httpClient.get<Penguin>(`api/penguin/${taxonKey}`)
  }

  /**
   * Send request to apply changes to given penguin.
   *
   * @param penguin penguin changes
   * @param taxonKey taxon key
   */
  putPenguin(penguin: NewPenguin, taxonKey: string): Observable<any> {
    return this.httpClient.put<NewPenguin>(`api/penguin/${taxonKey}`, penguin);
  }

  getImagesInfo(): Observable<ImagesInfo> {
    return this.httpClient.get<ImagesInfo>('api/images');
  }

}

File: penguin-service/src/main/java/pl/swosek/sample/penguin/data/service/api/PenguinService.java
package pl.swosek.sample.penguin.data.service.api;

import pl.swosek.sample.penguin.data.repository.entity.Penguin;

import java.util.List;
import java.util.Optional;

/**
 * Service for penguins related operations.
 */
public interface PenguinService {

    /**
     * Fetches all penguins.
     */
    List<Penguin> findAllPenguins();

    /**
     * Fetches potential penguin with given taxon key.
     */
    Optional<Penguin> findPenguin(String taxonKey);

}
